### PR TITLE
Added Spot-Q Algorithm

### PR DESCRIPTION
**Pull request by adit98**
Training is much slower on CPU (~8-10 it/s). May need to optimize further.

Now uses action mask instead of forward boolean value, all actions are allowed for selection, removed by action masking and the loss is propagated based on the spot-q algorithm.

Modified memory iterator to output the state and the allowed action mask (to prevent illegal actions during evaluation).

### CODE REVIEW

File: agent.py [hunk 1]
```diff
@@ -90,6 +115,38 @@ def learn(self, mem):
       m.view(-1).index_add_(0, (u + offset).view(-1), (pns_a * (b - l.float())).view(-1))  # m_u = m_u + p(s_t+n, a*)(b - l)
 
     loss = -torch.sum(m * log_ps_a, 1)  # Cross-entropy loss (minimises DKL(m||p(s_t, a_t)))
+
+    with torch.no_grad():
+      if spotq:
+        # get original chosen actions (for spot q loss)
+        argmax_indices_ns = dns.sum(2).argmax(1)
```
Comment (ahundt) [inline agent.py:123]: what happens if the argmax isn't a masked index? Is this for generally enabling spot-q or just enabling it in this case?

File: agent.py [hunk 2]
```diff
@@ -68,10 +84,19 @@ def learn(self, mem):
       # Calculate nth next state probabilities
       pns = self.online_net(next_states)  # Probabilities p(s_t+n, ·; θonline)
       dns = self.support.expand_as(pns) * pns  # Distribution d_t+n = (z, p(s_t+n, ·; θonline))
-      argmax_indices_ns = dns.sum(2).argmax(1)  # Perform argmax action selection using online network: argmax_a[(z, p(s_t+n, a; θonline))]
+
+      # Action Masking 
+      argmax_indices_masked = dns.sum(2)
```
Comment (ahundt) [inline agent.py]: Can you just take the regular argmax here too without the mask? Then you might be able to get away whithout a second forward pass.

File: agent.py [hunk 3]
```diff
@@ -90,6 +115,38 @@ def learn(self, mem):
       m.view(-1).index_add_(0, (u + offset).view(-1), (pns_a * (b - l.float())).view(-1))  # m_u = m_u + p(s_t+n, a*)(b - l)
 
     loss = -torch.sum(m * log_ps_a, 1)  # Cross-entropy loss (minimises DKL(m||p(s_t, a_t)))
+
+    with torch.no_grad():
```
Comment (ahundt) [inline agent.py]: I didn't have to really do a second forward pass like this.

File: agent.py [hunk 4]
```diff
@@ -90,6 +115,38 @@ def learn(self, mem):
...
+        u[(l < (self.atoms - 1)) * (l == u)] += 1
+
+        # Distribute probability of Tz
+        m = states.new_zeros(self.batch_size, self.atoms)
+        offset = torch.linspace(0, ((self.batch_size - 1) * self.atoms), self.batch_size).unsqueeze(1).expand(self.batch_size, self.atoms).to(actions)
+        m.view(-1).index_add_(0, (l + offset).view(-1), (pns_a_spotq * (u.float() - b)).view(-1))  # m_l = m_l + p(s_t+n, a*)(u - b)
+        m.view(-1).index_add_(0, (u + offset).view(-1), (pns_a_spotq * (b - l.float())).view(-1))  # m_u = m_u + p(s_t+n, a*)(b - l)
+
+        spotq_loss = m * log_ps_a
+        spotq_loss[~extra_term_inds] = 0
```
Comment (ahundt) [inline agent.py:135]: why is the loss being assigned to 0 here? that seems strange, it is the reward we need to be zero

File: agent.py (hunk 3 above)
Comment (ahundt) [inline agent.py]: Can you explain why this extra no grad forward pass is being done? the forward pass above has gradients.

**Review by ahundt — CHANGES_REQUESTED** [review]
Integrating spot-q here seems to be a bit more complicated than in the original on-robot algorithm. I've sent a link to the code which handles it on the robot.
